train.py

```diff
@@ -0,0 +1,245 @@
+import os
+import math
+
+import hydra
+import torch
+import wandb
+from tqdm import tqdm
+import torch.optim as optim
+import torch.nn.functional as F
+from omegaconf import OmegaConf
+import torchmetrics.functional as M
+from huggingface_hub import upload_file
+from torch.utils.data import Subset, DataLoader
+
+from data.dataset import MidiDataset
+from models.pitch_encoder import PitchEncoder
+from models.velocity_time_encoder import VelocityTimeEncoder
+
+# from models.reverse_diffusion import Unet
+# from ecg_segmentation_dataset import ECGDataset
+# from models.forward_diffusion import ForwardDiffusion
+
+
+def makedir_if_not_exists(dir: str):
+    if not os.path.exists(dir):
+        os.makedirs(dir)
+
+
+def preprocess_dataset(dataset_name: str, batch_size: int, num_workers: int, *, overfit_single_batch: bool = False):
+    train_ds = MidiDataset(dataset_name, split="train")
+    val_ds = MidiDataset(dataset_name, split="validation")
+    test_ds = MidiDataset(dataset_name, split="test")
+
+    if overfit_single_batch:
+        train_ds = Subset(train_ds, indices=range(batch_size))
+        val_ds = Subset(val_ds, indices=range(batch_size))
+        test_ds = Subset(test_ds, indices=range(batch_size))
+
+    # dataloaders
+    train_dataloader = DataLoader(train_ds, batch_size=batch_size, num_workers=num_workers)
```

Comment:
```suggestion
    train_dataloader = DataLoader(train_ds, batch_size=batch_size, num_workers=num_workers, shuffle=True)
```

I think it's necessary to shuffle the train split - otherwise batches will be coming in creation order - mostly from the same piece and in sequence